File: android/src/main/java/com/navigine/naviginesdk/NaviginePolylineMapObject.java
package com.navigine.naviginesdk;

import android.content.Context;

import androidx.annotation.NonNull;

import com.navigine.idl.java.PolylineMapObject;

import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NaviginePolylineMapObject implements MethodChannel.MethodCallHandler {
  private final int id;
  private final Context context;
  private final PolylineMapObject mapObject;
  private final MethodChannel methodChannel;

  public NaviginePolylineMapObject (
    int id,
    Context context,
    PolylineMapObject mapObject,
    BinaryMessenger messenger
  ) {
    this.id = id;
    this.mapObject = mapObject;
    this.context = context;

    methodChannel = new MethodChannel(messenger, "navigine_sdk/navigine_polyline_map_object_" + id);
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "setPolyLine":
        result.success(setPolyLine(call));
        break;
      case "setWidth":
        result.success(setWidth(call));
        break;
      case "setColor":
        result.success(setColor(call));
        break;
      case "getType":
        result.success(getType(call));
        break;
      case "setVisible":
        result.success(setVisible(call));
        break;
      case "setInteractive":
        result.success(setInteractive(call));
        break;
      case "setStyle":
        result.success(setStyle(call));
        break;
      case "setData":
        setData(call);
        result.success(null);
        break;
      case "getData":
        result.success(getData(call));
        break;
      case "setTitle":
        result.success(setTitle(call));
        break;


      default:
        result.notImplemented();
        break;
    }
  }

  public PolylineMapObject getMapObject() {
    return mapObject;
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setPolyLine(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return mapObject.setPolyLine(Utils.locationPolylineFromJson((Map<String, Object>) params.get("locationPolyline")));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setColor(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Double red = (Double) params.get("red");
    Double green = (Double) params.get("green");
    Double blue = (Double) params.get("blue");
    Double alpha = (Double) params.get("alpha");
    return mapObject.setColor(red.floatValue(), green.floatValue(), blue.floatValue(), alpha.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setWidth(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Double width = (Double) params.get("width");
    return mapObject.setWidth(width.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int getType(MethodCall call) {
    return mapObject.getType().ordinal();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setVisible(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean visible = (Boolean) params.get("visible");
    return mapObject.setVisible(visible);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setInteractive(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean interactive = (Boolean) params.get("interactive");
    return mapObject.setInteractive(interactive);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setStyle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String style = (String) params.get("style");
    return mapObject.setStyle(style);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setData(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    byte[] data = (byte[]) params.get("style");
    mapObject.setData(data);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public byte[] getData(MethodCall call) {
    return mapObject.getData();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setTitle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String title = (String) params.get("title");
    return mapObject.setTitle(title);
  }
}

File: ios/Classes/NaviginePolylineMapObject.swift
import Foundation
import Navigine

public class NaviginePolylineMapObject: NSObject {
    private var id: Int32
    private var mapObject: NCPolylineMapObject
    private let methodChannel: FlutterMethodChannel!
    private let pluginRegistrar: FlutterPluginRegistrar!

    public required init(
        id: Int32,
        mapObject: NCPolylineMapObject,
        registrar: FlutterPluginRegistrar
    ) {
        self.id = id
        self.mapObject = mapObject

        methodChannel = FlutterMethodChannel(
            name: "navigine_sdk/navigine_polyline_map_object_\(id)",
            binaryMessenger: registrar.messenger()
        )

        self.pluginRegistrar = registrar

        super.init()

        weak var weakSelf = self
        self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setPolyLine":
            result(setPolyLine(call))
        case "setColor":
            result(setColor(call))
        case "setWidth":
            result(setWidth(call))
        case "getType":
            result(getType(call))
        case "setVisible":
            result(setVisible(call))
        case "setInteractive":
            result(setInteractive(call))
        case "setStyle":
            result(setStyle(call))
        case "setData":
            setData(call)
            result(nil)
        case "getData":
            result(getData(call))
        case "setTitle":
            result(setTitle(call))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func getMapObject() -> NCPolylineMapObject {
        return mapObject
    }

    public func setPolyLine(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let locationPoint = Utils.locationPolylineFromJson(params["locationPolyline"] as! [String : Any])

        return mapObject.setPolyLine(locationPoint)
    }

    public func setColor(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let red = params["red"] as! NSNumber
        let green = params["green"] as! NSNumber
        let blue = params["blue"] as! NSNumber
        let alpha = params["alpha"] as! NSNumber

        return mapObject.setColor(red.floatValue, green: green.floatValue, blue: blue.floatValue, alpha: alpha.floatValue)
    }

    public func getType(_ call: FlutterMethodCall) -> Int {
        return mapObject.getType().rawValue
    }

    public func setWidth(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let width = (params["width"] as! NSNumber)

        return mapObject.setWidth(width.floatValue)
    }

    public func setVisible(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let visible = (params["visible"] as! NSNumber)

        return mapObject.setVisible(visible.boolValue)
    }

    public func setInteractive(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let interactive = (params["interactive"] as! NSNumber)

        return mapObject.setInteractive(interactive.boolValue)
    }

    public func setStyle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let style = (params["style"] as! String)

        return mapObject.setStyle(style)
    }

    public func setData(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let data = (params["data"] as! FlutterStandardTypedData).data

        return mapObject.setData(data)
    }

    public func getData(_ call: FlutterMethodCall) -> Data {
        return mapObject.getData()
    }

    public func setTitle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let title = (params["title"] as! String)

        return mapObject.setTitle(title)
    }

    private func getPolylineImage(_ image: [String: Any]) -> UIImage {
        let type = image["type"] as! String
        let defaultImage = UIImage()

        if (type == "fromAssetImage") {
          let assetName = self.pluginRegistrar.lookupKey(forAsset: image["assetName"] as! String)

          return UIImage(named: assetName) ?? defaultImage
        }

        if (type == "fromBytes") {
          let imageData = (image["rawImageData"] as! FlutterStandardTypedData).data

          return UIImage(data: imageData) ?? defaultImage
        }

        return defaultImage
      }
}
